### ISSUE TITLE
Misleading error on missing semicolon after return

### ISSUE THREAD

**Issue by cfhammill** [stan-dev/stan#2533]
#### Summary:
Functions defined with a return and no terminating semicolon report a nonsensical error.

#### Description:
See Above

#### Reproducible Steps:

file: demo.stan

```
functions {
  real bad_func(){
    return 5
  }
}
```

R:

```
rstan::stan_model("demo.stan")
```

#### Current Output:

```
SYNTAX ERROR, MESSAGE(S) FROM PARSER:

variable "return" does not exist.
  error in 'model6e356df3d68_demo' at line 3, column 11
  -------------------------------------------------
     1: functions {
     2:   real demo(){
     3:     return 5
                  ^
     4:   }
  -------------------------------------------------
```

#### Expected Output:

```
error in 'model6e356df3d68_demo' at line 3, column 11
  -------------------------------------------------
     1: functions {
     2:   real demo(){
     3:     return 5
     4:   }
           ^
  -------------------------------------------------
PARSER EXPECTED: ";"
```

#### Current Version:
v2.17.2

**Comment by bob-carpenter** [comment]
This should definitely be fixed.  It can be combined with the previous error message issue #2526 to cut down on testing and reviewing, since both should be straightforward.  

The easiest fix would be to put the `void_return_statement` before `return_statement` as disjuncts, then have the `return_statement` commit after seeing the literal "return" (and no-follow) with `>` (rather than `>>`).  Then add a second disjunct that fails right away (it'll still need to be of type `expression`), e.g.,

```
return_statement_r
  %= (lit("return") >> no_skip[!char_("a-zA-Z0-9_")])
     > (expression_g(_r1)
         | (eps[return_error_msg_f(_pass, boost::phoenix::ref(error_msgs_))]
            > expression_g(_r1))
     >> lit(';') [validate_return_allowed_f(_r1, _pass,
                                     boost::phoenix::ref(error_msgs_))];
```

The `return_error_msg_f` function just reports the error and sets `_pass = false`, at which piont the whole return statement fails.

**Comment by mitzimorris** [comment]
Implemented per suggestion, here's an example of the parser output:

```
Non-void function must return expression of specified return type.
  error in 'utility-stub.stan' at line 4, column 11
  -------------------------------------------------
     2:   real bar(real x) {
     3:     real y = x + 2.0;
     4:     return
                  ^
     5:   }
  -------------------------------------------------

PARSER EXPECTED: <alternative><expression><expect><eps><expression>
```

the parser reports the disjunct - confusing?  it does flag the error properly.

**Comment by mitzimorris** [comment]
Test for return statement missing semicolon:

```
SYNTAX ERROR, MESSAGE(S) FROM PARSER:

  error in 'utility-stub.stan' at line 5, column 3
  -------------------------------------------------
     3:     real y = x + 2.0;
     4:     return y
     5:   }
          ^
     6: }
  -------------------------------------------------

PARSER EXPECTED: ";"

```

**Comment by mitzimorris** [comment]
fixed by PR #2535 
closing
